### PR TITLE
Forward hook bluefog

### PR DESCRIPTION
**Pull request by Bluefog-Lib**
Before we trigger the communication after the backward gradient computation is finished in each layer. Actually we can put it at the end of forward computation:

w_{i+1} = neighbor_allreduce(w_{i}) - lr * gradient(w_{i})

### CODE REVIEW

File: bluefog/torch/optimizers.py [hunk 1]
```diff
@@ -275,62 +269,62 @@ def __init__(self, params, named_parameters):
...
-        for param_group in self.param_groups:
-            for p in param_group["params"]:
-                if p.requires_grad:
-                    p.register_hook(self._make_hook(p))
+    def _register_hook(self):
+        for parent_name, layer in self._model._modules.items():
+            layer.register_forward_hook(self._make_hook(parent_name))
 
-    def _make_hook(self, p):
+    def _make_hook(self):
```
Comment (hanbinhu) [inline bluefog/torch/optimizers.py:278]: 这个接口不对吧，和276行的不一致

Comment (ybc1991) [reply bluefog/torch/optimizers.py]: 你说的对
